unified-runtime/source/adapters/opencl/kernel.hpp

```diff
@@ -0,0 +1,56 @@
+//===--------- kernel.hpp - OpenCL Adapter ---------------------------===//
+//
+// Copyright (C) 2025 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "common.hpp"
+#include "context.hpp"
+#include "program.hpp"
+
+#include <vector>
+
+struct ur_kernel_handle_t_ {
+  using native_type = cl_kernel;
+  native_type CLKernel;
+  ur_program_handle_t Program;
+  ur_context_handle_t Context;
+  std::atomic<uint32_t> RefCount = 0;
+  bool IsNativeHandleOwned = true;
+
+  ur_kernel_handle_t_(native_type Kernel, ur_program_handle_t Program,
+                      ur_context_handle_t Context)
+      : CLKernel(Kernel), Program(Program), Context(Context) {
+    RefCount = 1;
+    if (Program) {
```

Comment:
I think a similar thing to the event-queue relationship applies here

Reply:
There's actually no way for this constructor to be called where Program is null, and ditto for the device branch below.

Reply:
I think null program might need to keep working for sycl reasons

Reply:
urKernelCreateFromNativeHandle, if given a null pointer as the program, will silently create (and validate) a handle to the program.

Reply:
so it does